src/stats/kde.nim

```diff
@@ -0,0 +1,209 @@
+# Copyright 2020 the Arraymancer contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import ../tensor/tensor,
+       ./distributions
+
+import std / [math, strutils]
+when (NimMajor, NimMinor, NimPatch) >= (1, 3, 0):
+  export nimIdentNormalize # for parseEnum
+
+type
+  KernelKind* = enum
+    knCustom = "custom"
+    knBox = "box"
+    knTriangular = "triangular"
+    knTrig = "trigonometric"
+    knEpanechnikov = "epanechnikov"
+    knGauss = "gauss"
+
+  KernelFunc* = proc(x, x_i, bw: float): float {.inline.}
```

Comment:
Does inline actually work for a callback?

Reply:
Oh, good point. I wasn't thinking about callbacks when I defined the type. But it is one of course. I would be surprised if inlining that worked, but I can check the generated code.

Reply:
Decided to have the actual `kde` implementation proc to take `static KernelFunc` and clean up the convenience `kde` implementation by using a template. Works well and should allow inlining (at least to "strip" the kernel wrapper of the underlying distribution proc).